Resolves issue 621 - Verify Address for exported Xpub Option

fedebuyito opened the pull request:
## Description

Because workflow of exportpubkey would must to be securely closed verifying addresses obtained on wallet software (Sparrow, for instance), issue reported #621 is conceptual consistent. 


This pull request is categorized as a:

- [X] New feature
- [ ] Bug fix
- [ ] Code refactor
- [ ] Documentation
- [X] Other (export XPUB address from seed and verify addresses workflow)

## Checklist

- [X] I’ve run `pytest` and made sure all unit tests pass before sumbitting the PR. They was updated, too.

If you modified or added functionality/workflow, did you add new unit tests?

- [ ] No, I’m a fool
- [ ] Yes
- [X] N/A

I have tested this PR on the following platforms/os:

- [ ] Raspberry Pi OS [Manual Build](https://github.com/SeedSigner/seedsigner/blob/dev/docs/manual_installation.md)
- [ ] [SeedSigner OS](https://github.com/SeedSigner/seedsigner-os) on a Pi0/Pi0W board
- [X] Other (enteropositivo emulator), new workflow was probed with sparrow wallet for segwit, nested and legacy scripts (taproot was not implemented)

Fixes #621 
Note: Keep your changes limited in scope; if you uncover other issues or improvements along the way, ideally submit those as a separate PR. The more complicated the PR the harder to review, test, and merge.

fedebuyito committed:
Update tools_screens.py

Changing KEY_BACKSPACE size on ToolsDiceEntropy and ToolsCoinFlip screens

fedebuyito committed:
Update tools_screens.py

Missed import sentence for commands added on last commit

fedebuyito committed:
Revert "Update tools_screens.py"

This reverts commit 7c26811239760159437cfd8146d489291a45e4ea.

fedebuyito committed:
Revert "Update tools_screens.py"

This reverts commit e8b0e35b27bedbbab324a82d3d61d5be350c3fa7.

fedebuyito committed:
Update seed_views.py

On class SeedExportXpubQRDisplayView(View): Replaces "Destination" return call from MainMenuView to ScanAddressView. Rest of changes are for passing seed_num between class calling.

fedebuyito committed:
Update scan_views.py

Changes for passing seed_num between class calling (for ScanAddressView class).

fedebuyito committed:
Update test_flows_seed.py

Test for seed views workflow affected.

fedebuyito commented:
Hi. Here, local pytest runned on 3.10 / 3.12 and video capture of workflow on Seedsigner Emulator (with loop backed camera for Sparrow interaction) for signet/segwit addresses:

![image](https://github.com/user-attachments/assets/753400c1-05e0-4361-b375-720302c00141)

![image](https://github.com/user-attachments/assets/03e01f5a-cfc9-443e-b5c0-7996005dddfa)

![2024-12-19 23-46-28](https://github.com/user-attachments/assets/01e544ff-21a5-4c35-be29-d6dd014d8ce6)

Tested with same results on my pi0-seedsigner device, too.

Best Regards. -

kdmukai commented:
Very cool.

But I would suggest an intermediate screen. We have to explain what the second camera live preview is expecting to scan.

* xpub QR export.
* Intermediate screen prompt: "Verify New Wallet?"; brief 1 sentence explanation of why / how(?); buttons:
  * [QR icon] Verify an address
  * Done
* Live camera preview to scan addr to verify.

fedebuyito commented:
Thanks for your feedback, Keith. I will try, that makes sense.

fedebuyito committed:
Update seed_views.py

Added SeedExportXpubQRAskVerifyAddView(View) class for intermediate screen asking for verifying address or done action after exporting Xpub

fedebuyito committed:
Update test_flows_seed.py

Added FlowStep(seed_views.SeedExportXpubQRAskVerifyAddView, screen_return_value=0) to seed flow tests.

fedebuyito commented:
Do you think this way would be fit? (I updated tests too, because of the new screen)

![image](https://github.com/user-attachments/assets/7b8683a4-5598-436a-a3e3-b93831569ae1)

fedebuyito committed:
Update screen.py

Added AdviceScreen class

fedebuyito committed:
Update scan_views.py

Resolves conflicts from previous merger

fedebuyito committed:
Update seed_views.py

Resolves conflicts from previous merger

fedebuyito committed:
Update scan_views.py

Resolves conflicts from previous merger

fedebuyito committed:
Update seed_views.py

Resolves conflicts from previous merger

fedebuyito committed:
Update screen.py

Resolves conflicts from previous merger

fedebuyito committed:
Update screen.py

Added AdviceScreen class.

fedebuyito committed:
Merge branch 'dev' into resolves_issue_621

fedebuyito committed:
Update scan_views.py

Resolves conflicts from previous merger

fedebuyito committed:
Update seed_views.py

Resolves conflicts from previous merger

fedebuyito committed:
Update test_flows_seed.py

Added FlowStep(seed_views.SeedExportXpubQRAskVerifyAddView, screen_return_value=0) to seed flow tests.

fedebuyito committed:
Update generator.py

Added  new Screenshot for SeedExportXpubQRAskVerifyAddView

fedebuyito commented:
- Code adapted in order to resolve conflicts from last massive merge (PR [#620](tg://search_hashtag?hashtag=620) ), including workflow test.
- Revised view text and icon for better understanding.
- Tested on local for 3.12 and 3.10, and pi0 manual build.
- Added new view to screenshot generator (tests passed, too).
- Did not try with multi language because is out of scope of this issue

![image](https://github.com/user-attachments/assets/ffbcb1d4-889a-45e1-bd6e-1b1c42119451)

![image](https://github.com/user-attachments/assets/af282890-7f07-4d58-8811-c8e8477d7f51)
![image](https://github.com/user-attachments/assets/9c5d9671-3014-4e65-81b2-070cbb35913d)

jdlcdl commented:
Reviewed and tried this pr at 2047506

This is nice functionality. I like that it helps teach users to verify their watch-only wallet setup.

* Perhaps improved wording on the new Advice screen:
  A better title might be "Wallet Verification".  Text might be "Scan a wallet address to verify the exported xpub?".  Button might be "Scan an address".
   ...once done, to make suggestions to translators/transifex for new strings in Spanish.
* Since multisig cannot work with a single xpub (though this pr does check and succeed address verifications), it might only be relevant to, and therefore available, after exporting single-sig xpubs.  It's not expected that the user would have access to the full descriptor until all bip32 xpubs have been exported... at that time, an advanced multisig user should know to verify their wallet setup with a complete wallet descriptor.

fedebuyito commented:
Hi @jdlcdl . Thanks for reviewing and comments. 

Sounds good that wording improvement. Note the headline with text "Check new wallet", below icon, is it useful? would you change it? 

![image](https://github.com/user-attachments/assets/f2ef7e22-9ed4-41a5-b98f-958e20f91427)
![image](https://github.com/user-attachments/assets/9c9ad635-0a57-4b8b-8949-93d72e47befd)

About this, 
>     * Since multisig cannot work with a single xpub (though this pr does check and succeed address verifications), it might only be relevant to, and therefore available, after exporting single-sig xpubs.  It's not expected that the user would have access to the full descriptor until all bip32 xpubs have been exported... at that time, an advanced multisig user should know to verify their wallet setup with a complete wallet descriptor.
I agree, and think I'm almost there.

jdlcdl commented:
Just noticing that this pr raises an error if an address is scanned before any seed has been loaded.  What is expected instead is to land at the "load a seed" view so that user can load a seed in various ways.

Error looks like:
```
System Error
(!)
AttributeError
scan_views.py, 136, in run
'ScanView' object has no attribute 'seed_num'
(I Understand)
```

Will leave a marker in code comment if I see it.

Otherwise, things are a moving target lately, we'll catch all the bugs before merge/release.
Have a safe and wonderful end-of-year celebration @fedebuyito.  All the best for 2025!

jdlcdl commented on src/seedsigner/views/scan_views.py:
```diff
@@ -133,6 +133,8 @@ def run(self):
                 from seedsigner.views.seed_views import AddressVerificationStartView
                 address = self.decoder.get_address()
                 (script_type, network) = self.decoder.get_address_type()
+                seed_num = self.seed_num
```
Cannot always be assumed that object has .seed_num, sometimes user will scan an address first, afterwards should be prompted to load a seed.

fedebuyito replied:
Fixed on [1e63527](https://github.com/SeedSigner/seedsigner/pull/629/commits/1e6352739d8610f7732ac58d92ee37946cee393d) commit.

fedebuyito committed:
Update scan_views.py

only includes "seed_num" when this exists.

fedebuyito committed:
Update seed_views.py

Only ask for verification xpub address option when SINGLE_SIG

fedebuyito committed:
Update seed_views.py

Solves view wording

fedebuyito committed:
Update test_flows_seed.py

Fixes test flows regarding ask xpub address verification only when SINGLE_SIG

fedebuyito committed:
Update seed_views.py

Resolves missing member declaration

fedebuyito commented:
> Just noticing that this pr raises an error if an address is scanned before any seed has been loaded. What is expected instead is to land at the "load a seed" view so that user can load a seed in various ways.
> 
> Error looks like:
> 
> ```
> System Error
> (!)
> AttributeError
> scan_views.py, 136, in run
> 'ScanView' object has no attribute 'seed_num'
> (I Understand)
> ```
> 
> Will leave a marker in code comment if I see it.
> 
> Otherwise, things are a moving target lately, we'll catch all the bugs before merge/release. Have a safe and wonderful end-of-year celebration @fedebuyito. All the best for 2025!

Thank you! I was not realized that there are two ways to scan an address before load a seed (Tools/Verify and SCAN directly, last I did catch it, curiously the most obvious). I logged the fix on marked commented code.

I have solved you suggested about not to ask xpub address verification when multi-signature on workflow, updating workflow test, too (all pytested and emulator tested). On screen wording I left without that yellow words but I read your opinion, of course, so you can tell me what you think is better.

@jdlcdl Nice to meet you on passed 2024 and thanks again for your kindly reviews and all the best for you too right now! 2025 will rock!

fedebuyito committed:
Merge branch 'dev' into resolves_issue_621

alvroble commented on src/seedsigner/gui/screens/screen.py:
```diff
@@ -1024,6 +1024,14 @@ class WarningScreen(WarningEdgesMixin, LargeIconStatusScreen):
 
 
 
+@dataclass
+class AdviceScreen(WarningEdgesMixin, LargeIconStatusScreen):
+    title: str = _mft("Advice")
+    status_icon_name: str = SeedSignerIconConstants.WARNING
+    status_color: str = "yellow"
+
+
+
```
I would suggest using the WarningScreen here just for the sake of following UX guidelines by @easyuxd

alvroble commented on src/seedsigner/views/scan_views.py:
```diff
@@ -142,6 +142,7 @@ def run(self):
                         "address": address,
                         "script_type": script_type,
                         "network": network,
+                        **({"seed_num": self.seed_num} if "seed_num" in self.__dict__ else {})
```
I've not seen any other use of this dynamic dictionary format (`**`) in the entire codebase. It makes me wonder if it's the best way to implement this. It works, though.

alvroble commented on src/seedsigner/views/seed_views.py:
```diff
@@ -115,13 +117,16 @@ def run(self):
-            is_button_text_centered=False,
-            button_data=button_data,
-        )
+        if self.seed_num==None:
```
```suggestion
        if self.seed_num == None:
```

alvroble commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1007,7 +1014,45 @@ def run(self):
+        # Because we have an explicit "DONE" button, we disable "BACK" to keep the
+        # routing options sane.
+        selected_menu_num = self.run_screen(
+            AdviceScreen,
```
Tested here with `WarningScreen` and the behavior is the same.

alvroble commented on tests/test_flows_seed.py:
```diff
@@ -361,7 +364,7 @@ def test_export_xpub_skip_non_option_flow(self):
                 FlowStep(seed_views.SeedExportXpubWarningView, screen_return_value=0),
                 FlowStep(seed_views.SeedExportXpubDetailsView, screen_return_value=0),
-                FlowStep(seed_views.SeedExportXpubQRDisplayView, screen_return_value=0),
+                FlowStep(seed_views.SeedExportXpubQRDisplayView, screen_return_value=0),             
```
This is probably an unintended modification of the line spacing.

alvroble commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1007,7 +1014,45 @@ def run(self):
+
+
+
+class SeedExportXpubQRAskVerifyAddView(View):
```
Here I'd use `SeedExportXpubQRAskVerifyAddressView`. It's longer but less confusing.

alvroble commented on tests/test_flows_seed.py:
```diff
@@ -199,6 +199,7 @@ def flowtest_standard_xpub(sig_tuple, script_tuple, coord_tuple):
                     FlowStep(seed_views.SeedExportXpubWarningView, screen_return_value=0),
                     FlowStep(seed_views.SeedExportXpubDetailsView, screen_return_value=0),
                     FlowStep(seed_views.SeedExportXpubQRDisplayView, screen_return_value=0),
+                    *( [FlowStep(seed_views.SeedExportXpubQRAskVerifyAddView, screen_return_value=0), FlowStep(scan_views.ScanAddressView, screen_return_value=0) ] if sig_selection == seed_views.SeedExportXpubSigTypeView.SINGLE_SIG else [] ),
```
I've not seen any other use of this dynamic list format (*) in the entire codebase. It makes me wonder if it's the best way to implement this. It works, though.

alvroble commented:
Tested ACK as of https://github.com/SeedSigner/seedsigner/pull/629/commits/c69f3c24481eb4a97b95bd63b500316049d11b3a on Raspberry Pi OS manual Build and working fine. It's a neat feature, given that the code nits are solved (specially I think we don't need the `AdviceScreen`).

newtonick commented:
@fedebuyito 

Sorry this PR has taken so long to get a proper look at.

As it stands, it isn't mergeable because it conflicts with the current dev branch and includes changes unrelated to the xpub address verification feature. Also the new UI strings like "Wallet Verification" are missing translation wrappers.

Do you still want to pursue this PR? I'm happy to circle back do a deep review if you do. If so, please rebase onto current dev and narrow the scope to just the xpub verification flow with proper tests. Fix a new of the items listed above. If there's no response in about a month, we'll close it for now.

fedebuyito commented:
Hi, @newtonick . Yes, sure. I take a try. Thank you very much.
